### PR TITLE
Support get a product's release' all tasks from Product Pages

### PR DESCRIPTION
**Pull request by zhanglu1122**
1. Created an client to connect to PP for release api
2. Updated README on some command and links
3. Rename customize logging class, otherwise it confused with internal logging class
4. Adding Tests

JIRA: RHELWF-10975

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -18,15 +18,17 @@ opentelemetry-instrumentation-flask = "<1.0"
...
-pytest = {version = "^8.2.0", optional = true}
-pytest-cov = {version = "^5.0.0", optional = true}
+requests = "^2.32.0"
+requests-kerberos = "^0.14.0"
 
-[tool.poetry.extras]
-test = [
-    "pytest",
-    "pytest-cov",
-]
```
Comment (hluk) [inline pyproject.toml:29]: These are extra dependencies required by tests installed by `tox` (see `commands_pre` in `tox.ini`).

File: src/retasc/constants.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+"""
+Constants Values
+"""
+
+# Product Pages
+PP_REST = "https://pp.engineering.redhat.com/api/latest/"
+# rhel_av and rhscl will be discontinued
+SUPPORTED_PRODUCTES = ["rhel", "cs", "directory", "dts", "dotnet", "rhel_av", "rhscl"]
```
Comment (hluk) [inline src/retasc/constants.py:8]: Typo: `SUPPORTED_PRODUCTS`

File: src/retasc/constants.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+"""
+Constants Values
+"""
+
+# Product Pages
+PP_REST = "https://pp.engineering.redhat.com/api/latest/"
```
Comment (hluk) [inline src/retasc/constants.py:6]: The real URL should be set in our internal repository. This public repos should not include any internal stuff, probably not even the products listed below this line.

We can pass configuration in **environment variables** (see the section in `REAMDE.md`).

File: src/retasc/constants.py (hunk 2 above)
Comment (zhanglu1122) [reply src/retasc/constants.py:8]: ack

File: src/retasc/constants.py (hunk 3 above)
Comment (zhanglu1122) [reply src/retasc/constants.py:6]: ack

File: pyproject.toml (hunk 1 above)
Comment (zhanglu1122) [reply pyproject.toml:29]: okay, I thought they can be replaced by using [tool.poetry.group.dev.dependencies], I will add them back.

Comment (hluk) [reply pyproject.toml:29]: Installing dev dependencies might be OK too if there are only deps for unit tests. Not sure what `--dev-dependency` parameter does, for me it is not listed with `poetry install --help`.

File: pyproject.toml [hunk 4]
```diff
@@ -18,9 +18,15 @@ opentelemetry-instrumentation-flask = "^0.45b0"
...
-pytest-cov = {version = "^5.0.0", optional = true}
+requests = "^2.32.0"
+requests-kerberos = "^0.14.0"
+
+urllib3 = "^2.2.1"
+attrs = "^23.2.0"
+
+[tool.poetry.group.dev.dependencies]
+requests-mock = "^1.12.1"
+black = "^24.4.0"
```
Comment (hluk) [inline pyproject.toml:29]: `black` should not be need since we already use `ruff` to auto-format the code in `pre-commit`.

Comment (zhanglu1122) [reply pyproject.toml]: removed.

File: .tekton/retasc-pull-request.yaml [hunk 5]
```diff
@@ -232,7 +232,7 @@ spec:
         - name: name
           value: buildah
         - name: bundle
-          value: quay.io/redhat-appstudio-tekton-catalog/task-buildah:0.1@sha256:6b483e420a9abc7462dd0a4886a520bf44fb8d1cb558badf30a67af0d2fd21fc
+          value: quay.io/redhat-appstudio-tekton-catalog/task-buildah:0.1@sha256:102500165339bc08791775cf2c4dcae3dd4bde557a9009d44dc590ef66dde384
```
Comment (hluk) [inline .tekton/retasc-pull-request.yaml:235]: What's up with these changes? Feel free to merge any open PRs from Konflux separately (if they pass CI).

Comment (zhanglu1122) [reply .tekton/retasc-pull-request.yaml:235]: I feel this is not from me, maybe got from some rebase action

Comment (zhanglu1122) [reply .tekton/retasc-pull-request.yaml:235]: https://github.com/release-engineering/retasc/blob/main/.tekton/retasc-pull-request.yaml#L235, see here, it is same value

**Comment by zhanglu1122** [comment]
@hluk all tests passed, PTAL, thx

File: .tekton/retasc-pull-request.yaml (hunk 5 above)
Comment (hluk) [reply .tekton/retasc-pull-request.yaml]: I can still see the commit, but it seems to contain unrelated changes. Can you squash the commits?

File: src/retasc/product_pages_api.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        phase id 1000 is unsupported, `lt` is less than
+        """
+        opt = {
+            "product__shortname": product_shortname,
+            "fields": fields,
+        }
+        if active_only is True:
+            opt.update({"phase__lt": constants.UNSUPPORTED_PHASE_ID})
+        res = _no_auth_session.get(self.pp_release_url, params=opt)
+        return res.json()
```
Comment (hluk) [inline src/retasc/product_pages_api.py:42]:
```suggestion
        res.raise_for_status()
        return res.json()
```

File: src/retasc/product_pages_api.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        https://{pp_release_url}{release_short_name}/schedule-tasks?fields=a,b,c
+        The response have the following fields:name,,date_start,date_finish,flags
+        """
+        res = _no_auth_session.get(
+            f"{self.pp_release_url}{release_short_name}/schedule-tasks",
+            params={
+                "fields": fields,
+            },
+        )
+        return res.json()
```
Comment (hluk) [inline src/retasc/product_pages_api.py:61]:
```suggestion
        res.raise_for_status()
        return res.json()
```

File: src/retasc/product_pages_api.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        The response have the following fields:release_shortname,name,slug,date_start,date_finish
+        """
+        res = _no_auth_session.get(
+            f"{self.pp_release_url}{release_short_name}/schedule-tasks",
+            params={
+                flag_cond: flags,
+                "fields": fields,
+            },
+        )
+        return res.json()
```
Comment (hluk) [inline src/retasc/product_pages_api.py:86]:
```suggestion
        res.raise_for_status()
        return res.json()
```

File: src/retasc/product_pages_api.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        The response have the following fields:release_shortname,name,slug,date_start,date_finish
+        """
+        res = _no_auth_session.get(
+            f"{self.pp_release_url}{release_short_name}/schedule-tasks",
+            params={
+                "name__regex": condition,
+                "fields": fields,
+            },
+        )
+        return res.json()
```
Comment (hluk) [inline src/retasc/product_pages_api.py:107]:
```suggestion
        res.raise_for_status()
        return res.json()
```

File: src/retasc/product_pages_api.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+"""
+Production Pages API
+"""
+
+import os
+from functools import lru_cache
+
+from retasc import constants, requests_session
+
+_no_auth_session = requests_session.requests_session()
```
Comment (hluk) [inline src/retasc/product_pages_api.py:11]: Better to move this to `ProductPagesApi`. Ideally if this is parameter in constructor and does not indicate what kind of session it is in name:

```python
    def __init__(self, api_url: str, session: requests.Session):
        self.api_url = api_url
        self.session = session
```

File: tests/test_product_pages_api.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
+# SPDX-License-Identifier: GPL-3.0+
+import pytest
+import requests_mock
+
+from retasc.constants import SUPPORTED_PRODUCTS, UNSUPPORTED_PHASE_ID
+from retasc.product_pages_api import ProductPagesApi
+
+url = "https://pp.rh.com/"
```
Comment (hluk) [inline tests/test_product_pages_api.py:8]:
```suggestion
PP_URL = "https://product_pages.example.com/"
```

File: src/retasc/product_pages_api.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+            #         {"rhscl-3-8": [
+            #             {
+            #     "name": "Planning Phase",
+            #     "date_start": "2021-03-08",
+            #     "date_finish": "2021-05-10",
+            #     "flags": [
+            #         "phase"
+            #     ]
+            # },]}
+            # }
```
Comment (hluk) [inline src/retasc/product_pages_api.py:145]:
```suggestion
```

File: src/retasc/product_pages_api.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        https://{pp_release_url}?product__shortname={product_shortname}&phase__lt=1000&fields=shortname
+        Loop through the releases and get the tasks of it by rest api
+        https://{pp_release_url}{release_short_name}/schedule-tasks/fields=name,slug,date_start,date_finish
+        """
+
+        product_shortname = product_shortname.strip()
+        if product_shortname not in constants.SUPPORTED_PRODUCTS:
+            raise ValueError(
+                f"Product - {product_shortname} is not supported yet. The products we supported currently are {constants.SUPPORTED_PRODUCTS}"
+            )
```
Comment (hluk) [inline src/retasc/product_pages_api.py:123]: Why is this needed? Can we just support any product?

File: Containerfile [hunk 14]
```diff
@@ -10,6 +10,7 @@ RUN set -exo pipefail \
         --disablerepo=* \
         --enablerepo=fedora,updates \
         gcc \
+        krb5-devel \
```
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply Containerfile:13]: sure

File: src/retasc/constants.py [hunk 15]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: GPL-3.0+
+"""
+Constants Values
+"""
+
+# Product Pages
+# rhel_av and rhscl will be discontinued
+SUPPORTED_PRODUCTS = ["rhel", "cs", "directory", "dts", "dotnet", "rhel_av", "rhscl"]
+UNSUPPORTED_PHASE_ID = 1000
```
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/constants.py:9]: just some constants need to use in pp rest api call, where do you suggest to store? Did not get what do you mean internal repo and what should I do if I want to have a constants.py like the old way?

File: src/retasc/product_pages_api.py [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+_no_auth_session = requests_session.requests_session()
+
+
+class ProductPagesApi:
+    """
+    Product Pages API Client
+    """
+
+    def __init__(self):
+        self.pp_release_url = os.getenv("PP_URL", "PP_URL not set") + "releases/"
```
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:20]: sure, where to have all the get env, any suggestion?

File: src/retasc/product_pages_api.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    def releases_by_product_phase(
+        self,
+        product_shortname: str,
+        active_only: bool = True,
+        fields: str = "shortname,product_shortname,ga_date,phase_display",
+    ) -> list[dict]:
+        """
+        https://{pp_release_url}?product__shortname={product_shortname}&phase__lt=1000&fields=a,b,c
+        The response have the following fields:shortname,product_shortname,ga_date,phase_display
+        phase id 1000 is unsupported, `lt` is less than
```
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:33]: sure.

File: src/retasc/product_pages_api.py (hunk 6 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:42]: thx, updated for all response

File: src/retasc/product_pages_api.py (hunk 9 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:107]: ack

File: src/retasc/product_pages_api.py (hunk 10 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:11]: ack

Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:11]: Do you mean don't use `_no_auth_session = requests_session.requests_session()`?
Or 
```
def __init__(self, api_url: str, session: requests_session.requests_session()):
```

File: src/retasc/product_pages_api.py (hunk 12 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:145]: ack

File: tests/test_product_pages_api.py (hunk 11 above)
Comment (zhanglu1122) [reply tests/test_product_pages_api.py:8]: ack, deleted all the mock set env as well.

File: src/retasc/product_pages_api.py (hunk 13 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:123]: Not all, we only supported those(https://spaces.redhat.com/pages/viewpage.action?spaceKey=EXDSPRHEL&title=SP+RHEL+RelOps+Upcoming+work) and Some of the product will be sunset as well and can be removed later.
Only support what I have tested, not others, can be extend in future?

File: Containerfile (hunk 14 above)
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply Containerfile]: This is needed here, otherwise build container step will failed, https://console.redhat.com/preview/application-pipeline/workspaces/exd-sp-rhel-wf/applications/retasc/pipelineruns/retasc-on-pull-request-l9t4p/logs?task=build-container
It is at the same location as https://github.com/release-engineering/resultsdb/blob/480b63f262c7bc80e26c30fba42f2f7757a74917/Dockerfile#L13

File: .tekton/retasc-pull-request.yaml (hunk 5 above)
Comment (zhanglu1122) [reply .tekton/retasc-pull-request.yaml]: not sure how to do the git squash, in gitlab I normally do this action before merge in UI.

Comment (zhanglu1122) [reply .tekton/retasc-pull-request.yaml]: squashed

File: Containerfile (hunk 14 above)
[the remark this answers is not in the document]
Comment (hluk) [reply Containerfile]: I understand it is needed, but you are likely linking against krb5 libraries when poetry is building some Python packages, so `krb5-libs` must be installed in the final container image (see it in the resultsdb Dockerfile) otherwise the resulting container will have missing library dependencies.

File: src/retasc/product_pages_api.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (hluk) [reply src/retasc/product_pages_api.py:31]: yes

File: src/retasc/product_pages_api.py (hunk 13 above)
Comment (hluk) [reply src/retasc/product_pages_api.py:129]: But why limit the supported products here? We can remove/disallow any unsupported products in our internal configuration, right?

File: src/retasc/product_pages_api.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (hluk) [reply src/retasc/product_pages_api.py]: We can decide it later, it could be at the place where `ProductPagesApi` is instantiated.

File: Containerfile (hunk 14 above)
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply Containerfile]: I see what you mean now, I misunderstood your original comment. Fixing it now.

File: src/retasc/product_pages_api.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py]: sure, this has been updated

File: src/retasc/product_pages_api.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:31]: ack

File: src/retasc/product_pages_api.py (hunk 13 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:129]: sure, let me remove this

File: src/retasc/product_pages_api.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        product_shortname: str,
+        active_only: bool = True,
+        fields: str = "shortname,product_shortname,ga_date,phase_display",
+    ) -> list[dict]:
+        """
+        https://{pp_release_url}?product__shortname={product_shortname}&phase__lt=Unsupported&fields=a,b,c
+        The response have the following fields:shortname,product_shortname,ga_date,phase_display
+        phase id 1000 is unsupported, `lt` is less than
+        """
+        pp_release_url = self.api_url + "releases/"
```
Comment (hluk) [inline src/retasc/product_pages_api.py:33]: This assumes that `api_url` has trailing `/`. Maybe create a helper method:

```python
    def api_url_releases(self, id: int | None = None):
        return f"{self.api_url.rstrip('/')}/releases/{id or ''}"
```

File: src/retasc/constants.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (hluk) [reply src/retasc/constants.py]: `UNSUPPORTED_PHASE_ID` can be "Unsupported" and should be moved to product_pages_api.py.

`SUPPORTED_PRODUCTS` is probably not needed. Such values would be set from env variables or from a config file. Anyway, it is not a constant.

File: src/retasc/product_pages_api.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    """
+    Product Pages API Client
+    """
+
+    def __init__(self, api_url: str):
+        self.api_url = api_url
+        self.session = requests_session.requests_session()
+
+    @lru_cache(maxsize=None)
+    @staticmethod
```
Comment (hluk) [inline src/retasc/product_pages_api.py:21]: These are not static methods since they use class instance members (it takes `self` argument).

File: src/retasc/constants.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/constants.py:9]: this file is removed, no longer need SUPPORTED_PRODUCTS and moved UNSUPPORTED_PHASE_ID to pp api class

File: src/retasc/product_pages_api.py [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+        product_shortname = product_shortname.strip()
+        if product_shortname not in constants.SUPPORTED_PRODUCTS:
+            raise ValueError(
+                f"Product - {product_shortname} is not supported yet. The products we supported currently are {constants.SUPPORTED_PRODUCTS}"
+            )
+
+        releases_list = ProductPagesApi.releases_by_product_phase(
+            self, product_shortname
+        )
+        all_tasks = {}  # type: dict
```
Comment (hluk) [inline src/retasc/product_pages_api.py:128]:
```suggestion
        all_tasks: dict = {}
```

A fuller type hint (still incomplete) is `dict[str, dict[str, list[dict]]]` which indicates that the dict would be somewhat difficult to use. It might be better to wrap it in a smarter object.

File: src/retasc/product_pages_api.py [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    @staticmethod
+    def get_all_active_releases_tasks_for_product(self, product_shortname: str) -> dict:
+        """
+        Get all release names that are still supported by rest api:
+        https://{pp_release_url}?product__shortname={product_shortname}&phase__lt=Unsupported&fields=shortname
+        Loop through the releases and get the tasks of it by rest api
+        https://{pp_release_url}{release_short_name}/schedule-tasks/fields=name,slug,date_start,date_finish
+        """
+
+        releases_list = ProductPagesApi.releases_by_product_phase(
```
Comment (hluk) [inline src/retasc/product_pages_api.py:125]: Why not use `self.releases_by_product_phase()`?

File: src/retasc/product_pages_api.py [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+        The response have the following fields:shortname,product_shortname,ga_date,phase_display
+        phase id 1000 is unsupported, `lt` is less than
+        """
+        pp_release_url = self.api_url + "releases/"
+        opt = {
+            "product__shortname": product_shortname,
+            "fields": fields,
+        }
+        if active_only is True:
+            opt.update({"phase__lt": 1000})
```
Comment (hluk) [inline src/retasc/product_pages_api.py:39]:
```suggestion
            opt.update({"phase__lt": "Unsupported"})
```

File: tests/test_product_pages_api.py [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+            "product_shortname": test_prod,
+            "ga_date": "2021-11-15",
+            "phase_display": "Maintenance",
+        }
+    ]
+
+    # test all default parameters for ProductPagesApi.releases_by_product_phase, url should has all default fields
+    with requests_mock.Mocker() as m:
+        m.get(
+            f"{PP_URL}releases/?product__shortname={test_prod}&fields={default_fields}&phase__lt=1000",
```
Comment (hluk) [inline tests/test_product_pages_api.py:29]:
```suggestion
            f"{PP_URL}releases/?product__shortname={test_prod}&fields={default_fields}&phase__lt=Unsupported",
```

File: src/retasc/product_pages_api.py [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+    def releases_by_product_phase(
+        self,
+        product_shortname: str,
+        active_only: bool = True,
+        fields: str = "shortname,product_shortname,ga_date,phase_display",
+    ) -> list[dict]:
+        """
+        https://{pp_release_url}?product__shortname={product_shortname}&phase__lt=Unsupported&fields=a,b,c
+        The response have the following fields:shortname,product_shortname,ga_date,phase_display
+        phase id 1000 is unsupported, `lt` is less than
```
Comment (hluk) [inline src/retasc/product_pages_api.py:31]:
```suggestion
        phase__lt=Unsupported means "any supported phase" (these are less than Unsupported).
```

**Review by hluk — CHANGES_REQUESTED** [review]

File: src/retasc/product_pages_api.py [hunk 25]
```diff
@@ -0,0 +1,134 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+"""
+Production Pages API
+"""
+
+from functools import lru_cache
```
Comment (hluk) [inline src/retasc/product_pages_api.py:6]: Probably ok to use just `cache` (instead of `lru_cache(maxsize=None)`).

Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:6]: you mean 
```
from functools import cache

    @cache
    @staticmethod
    def releases_by_product_phase(
```
?

File: src/retasc/product_pages_api.py (hunk 24 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:31]: ack

File: src/retasc/product_pages_api.py (hunk 19 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:21]: sorry, forgot to remove, they were static but not anymore.

File: src/retasc/product_pages_api.py (hunk 18 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:33]: thanks, will update it

File: src/retasc/product_pages_api.py (hunk 20 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:128]: ack

File: src/retasc/product_pages_api.py (hunk 21 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:125]: was static methond, will update it

File: src/retasc/product_pages_api.py (hunk 22 above)
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:39]: ack, forgot to update in last commit

File: tests/test_product_pages_api.py (hunk 23 above)
Comment (zhanglu1122) [reply tests/test_product_pages_api.py:29]: ack

**Comment by zhanglu1122** [comment]
The linter failure is https://github.com/release-engineering/retasc/actions/runs/9611499774/job/26510188408?pr=12
```
Found 1 error in 1 file (checked 7 source files)
src/retasc/product_pages_api.py:42: error: Dict entry 0 has incompatible type "str": "int"; expected "str": "str"  [dict-item]
```
not sure what is that about, it complains about `res.raise_for_status()`

File: src/retasc/product_pages_api.py (hunk 25 above)
Comment (hluk) [reply src/retasc/product_pages_api.py]: yes

File: pyproject.toml [hunk 26]
```diff
@@ -18,15 +18,16 @@ opentelemetry-instrumentation-flask = "^0.45b0"
 opentelemetry-instrumentation-requests = "^0.45b0"
 opentelemetry-exporter-otlp-proto-http = "^1.24.0"
 
-# for tests
-pytest = {version = "^8.2.0", optional = true}
-pytest-cov = {version = "^5.0.0", optional = true}
+requests = "^2.32.0"
+requests-kerberos = "^0.14.0"
```
Comment (hluk) [inline pyproject.toml:22]: I don't see this being used anywhere.

File: src/retasc/requests_session.py [hunk 27]
```diff
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+"""
+Http Request client with retry and kerberos capabilities
```
Comment (hluk) [inline src/retasc/requests_session.py:3]:
```suggestion
HTTP request client with retry capability
```

File: src/retasc/product_pages_api.py [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+        release_short_name: str,
+        fields: str = "name,date_start,date_finish,flags",
+    ) -> list[dict]:
+        """
+        https://{pp_release_url}{release_short_name}/schedule-tasks?fields=a,b,c
+        The response have the following fields:name,,date_start,date_finish,flags
+        """
+
+        res = self.session.get(
+            f"{self.api_url_releases()}{release_short_name}/schedule-tasks",
```
Comment (hluk) [inline src/retasc/product_pages_api.py:58]:
```suggestion
            f"{self.api_url_releases(release_short_name)}/schedule-tasks",
```

But better to create a helper method:

```python
    def api_url_release_schedule_tasks(self, id: str) -> str:
        return f"{self.api_url_releases(id)}/schedule-tasks",
```

File: src/retasc/product_pages_api.py [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        https://{pp_release_url}{release_short_name}/schedule-tasks/?{flag_cond}={flags}&fields=a,b,c
+        Only one flag in condition use `flags_in=sp`
+        Multiple flags in OR condition use `flags_or__in=sp,rcm,releng`
+        Multiple flags in AND condition use `flags_and__in=sp,ga`
+        The response have the following fields:release_shortname,name,slug,date_start,date_finish
+        """
+
+        res = self.session.get(
+            f"{self.api_url_releases()}{release_short_name}/schedule-tasks",
```
Comment (hluk) [inline src/retasc/product_pages_api.py:83]:
```suggestion
            f"{self.api_url_releases(release_short_name)}/schedule-tasks",
```

File: src/retasc/product_pages_api.py [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+        condition: str,
+        fields: str = "release_shortname,name,slug,date_start,date_finish",
+    ) -> list[dict]:
+        """
+        https://{pp_release_url}{release_short_name}/schedule-tasks?name__regex={condition}&fields=a,b,c
+        The response have the following fields:release_shortname,name,slug,date_start,date_finish
+        """
+
+        res = self.session.get(
+            f"{self.api_url_releases()}{release_short_name}/schedule-tasks",
```
Comment (hluk) [inline src/retasc/product_pages_api.py:105]:
```suggestion
            f"{self.api_url_releases(release_short_name)}/schedule-tasks",
```

File: tests/test_product_pages_api.py [hunk 31]
```diff
@@ -0,0 +1,180 @@
...
+            "product_shortname": test_prod,
+            "ga_date": "2021-11-15",
+            "phase_display": "Maintenance",
+        }
+    ]
+
+    # test all default parameters for ProductPagesApi.releases_by_product_phase, url should has all default fields
+    with requests_mock.Mocker() as m:
+        m.get(
+            f"{PP_API.api_url_releases()}?product__shortname={test_prod}&fields={default_fields}&phase__lt=Unsupported",
```
Comment (hluk) [inline tests/test_product_pages_api.py:30]: This could be difficult to maintain. It would be better to have a helper test function to help with testing the parameters (maybe with `additional_matcher`: https://requests-mock.readthedocs.io/en/latest/matching.html#additional-matchers). Or check later if expected API endpoint was called with specific parameters (https://requests-mock.readthedocs.io/en/latest/history.html):

```python
    with requests_mock.Mocker() as m:
        m.get(PP_API.api_url_releases(), json=res)
        resp = PP_API.releases_by_product_phase(test_prod)
        assert len(m.request_history) == 1
        req = m.request_history[0]
        assert ("product__shortname", test_prod) in req.params
        # these are probably not important to test - probably ok to drop
        assert ("phase__lt", "Unsupported") in req.params
        assert ("fields", default_fields) in req.params
```

**Review by hluk — CHANGES_REQUESTED** [review]

File: src/retasc/product_pages_api.py [hunk 32]
```diff
@@ -0,0 +1,130 @@
...
+        self.session = requests_session.requests_session()
+
+    def api_url_releases(self, id: str | None = None) -> str:
+        return f"{self.api_url.rstrip('/')}/releases/{id or ''}"
+
+    @cache
+    def releases_by_product_phase(
+        self,
+        product_shortname: str,
+        active_only: bool = True,
```
[the remark this answers is not in the document]
Comment (zhanglu1122) [reply src/retasc/product_pages_api.py:27]: if we want to get all the release(even with unsupported releases). If drop this parameter the func name will be updated to `active_ releases_by_product_phase `, do you want to do change to this?

[the remark this answers is not in the document]
Comment (hluk) [reply src/retasc/product_pages_api.py:27]: Why would we want to process inactive releases? What `active=False` actually means? I would expect that there is no planned/upcoming schedule items. I don't think retasc would need to process these.

Either name `releases()` or `active_releases()` is fine with me.

File: Containerfile (hunk 14 above)
Comment (hluk) [inline Containerfile:13]: This and `krb5-libs` also should not needed anymore.

File: src/retasc/product_pages_api.py [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+        https://{pp_release_url}{release_short_name}/schedule-tasks/?{flag_cond}={flags}&{name__regex}={reg_query}&fields=a,b,c
+        """
+
+        url = self.api_url_release_schedule_tasks(release_short_name)
+        # fields has some default value, can be updated if kargs has it
+        addtional_query = {
+            "fields": "release_shortname,name,slug,date_start,date_finish"
+        }
+        for k, v in kargs.items():
+            addtional_query.update({k: v})
```
Comment (hluk) [inline src/retasc/product_pages_api.py:77]:
```suggestion
    def release_schedules_by_search_query(
        self, release_short_name: str, fields: str = "release_shortname,name,slug,date_start,date_finish", **kwargs
    ) -> list[dict]:
        """
        https://{api_url_release_schedule_tasks}/?{kwargs...}
        """

        url = self.api_url_release_schedule_tasks(release_short_name)
```
 
But this looks the same as `release_schedules()` except `fields` is different and support for additional kwargs.

Can you add support for the kwargs to `release_schedules()` and remove `release_schedules_by_search_query()` method?

File: src/retasc/product_pages_api.py [hunk 34]
```diff
@@ -0,0 +1,99 @@
...
+        """
+
+        releases_list = self.active_releases(product_shortname)
+        all_tasks: dict = {}
+        all_tasks[product_shortname] = {}
+        for r in releases_list:
+            r_name = r["shortname"]
+            r_tasks = self.release_schedules(r_name)
+            all_tasks[product_shortname][r_name] = r_tasks
+        return all_tasks
```
Comment (hluk) [inline src/retasc/product_pages_api.py:99]: This wraps the result unnecessarily in a dict containing just one item `product_shortname`.

```suggestion
        return {r["shortname"]: self.release_schedules(r["shortname"]) for r in releases_list}
```

File: tests/test_product_pages_api.py [hunk 35]
```diff
@@ -0,0 +1,141 @@
+# SPDX-License-Identifier: GPL-3.0+
+import requests_mock
+
+from retasc.product_pages_api import ProductPagesApi
+
+PP_URL = "https://product_pages.example.com"
+PP_API = ProductPagesApi(PP_URL)
```
Comment (hluk) [inline tests/test_product_pages_api.py:7]: This will not work well due to the caching. It would be better to create the object for each test method, e.g. using a pytest fixture:

```python
from pytest import fixture
...

@fixture
def pp_api():
    return ProductPagesApi(PP_URL)

...

def test_get_pp_url_info(pp_api):
    assert pp_api.api_url == PP_URL
```

File: tests/test_product_pages_api.py [hunk 36]
```diff
@@ -0,0 +1,141 @@
...
+test_prod = "rhscl"
+test_release = "rhscl-3-8"
+release_url = PP_API.api_url_releases()
+
+
+def request_assertion(prod_or_release, method_under_test, url, res, **kargs):
```
Comment (hluk) [inline tests/test_product_pages_api.py:13]: This does not help the tests to be easier to write or read. It would be much cleaner if the caller itself calls the method_under_test and `assert`.

File: tests/test_requests_session.py
Comment (hluk) [inline tests/test_requests_session.py:1]: Testing the RequestsSession is a bit overkill. It is mainly not our code and it should be indirectly covered by the other tests. I would drop this file.

**Review by hluk — CHANGES_REQUESTED** [review]

File: tests/test_product_pages_api.py [hunk 37]
```diff
@@ -0,0 +1,167 @@
+# SPDX-License-Identifier: GPL-3.0+
+import requests_mock
+from pytest import fixture
+
+from retasc.product_pages_api import ProductPagesApi
+
+PP_URL = "https://product_pages.example.com"
+PP_API = ProductPagesApi(PP_URL)
+test_prod = "rhscl"
+test_release = "rhscl-3-8"
```
Comment (hluk) [inline tests/test_product_pages_api.py:10]:
```suggestion
TEST_PRODUCT = "rhscl"
TEST_RELEASE = "rhscl-3-8"
```

File: tests/test_product_pages_api.py [hunk 38]
```diff
@@ -0,0 +1,167 @@
...
+            "shortname": "rhscl-3-8",
+            "product_shortname": test_prod,
+            "ga_date": "2021-11-15",
+            "phase_display": "Maintenance",
+        }
+    ]
+    default_fields = "shortname,product_shortname,ga_date,phase_display"
+    expected_url = f"{pp_api.api_url_releases()}?product__shortname={test_prod}&phase__lt=Unsupported&fields={default_fields}"
+    # test all default parameters for ProductPagesApi.active_releases, url should has all default fields
+    request_assertion(pp_api, "product", "active_releases", expected_url, test_res)
```
Comment (hluk) [inline tests/test_product_pages_api.py:53]:
```suggestion
    requests_mock.get(pp_api.api_url_releases(), json=test_res)
    resp = pp_api.active_releases(test_prod)
    assert len(requests_mock.request_history) == 1
    req = requests_mock.request_history[0]
    assert req.qs.get("product__shortname") == [test_prod]
    assert req.qs.get("fields") == [default_fields]
    assert resp == test_res
```

**Comment by zhanglu1122** [comment]
@hluk PTAL, I have made the suggested changes. Thanks

File: src/retasc/product_pages_api.py [hunk 39]
```diff
@@ -0,0 +1,81 @@
...
+    ) -> list[dict]:
+        """
+        https://{api_url_release_schedule_tasks}/?{kwargs...}
+        """
+        url = self.api_url_release_schedule_tasks(release_short_name)
+
+        # fields has some default value, can be updated if kwargs has it
+        addtional_query = {"fields": fields}
+        for k, v in kwargs.items():
+            addtional_query.update({k: v})
```
Comment (hluk) [inline src/retasc/product_pages_api.py:62]:
```suggestion
        addtional_query = {"fields": fields, **kwargs}
```

File: src/retasc/product_pages_api.py [hunk 40]
```diff
@@ -0,0 +1,81 @@
...
+
+        res = self.session.get(url, params=addtional_query)
+        res.raise_for_status()
+        return res.json()
+
+    @cache
+    def get_all_active_releases_tasks_for_product(self, product_shortname: str) -> dict:
```
Comment (hluk) [inline src/retasc/product_pages_api.py:69]: Can we rename this to `active_release_schedules_for_product` for consistency?

**Review by hluk — COMMENTED** [review]
Two minor comment, otherwise LGTM! 👍

**Review by hluk — APPROVED** [review]
LGTM!
